src/databricks/labs/ucx/hive_metastore/table_acl_migrate.py

```diff
@@ -0,0 +1,174 @@
+            # user_mapping = self._get_cluster_principal_mapping(cluster_spn.cluster_id)
+            logger.info(f"Applying UC ACL for cluster id {cluster_spn.cluster_id}")
+            for spn in cluster_spn.spn_info:
+                self._apply_uc_permission(external_locations)
+
+    def _get_principal_prefix(
+        self, spn_cluster_mapping: list[AzureServicePrincipalClusterMapping]
+    ) -> list[StoragePermissionMapping]:
+        set_service_principals = set[AzureServicePrincipalInfo]()
+        for spn_mapping in spn_cluster_mapping:
+            set_service_principals.update(spn_mapping.spn_info)
+        external_locations = []
+        permission_mappings = self._resource_permission.load()
+        for spn in set_service_principals:
+            for perm_mapping in permission_mappings:
+                if perm_mapping.client_id == spn.application_id and spn.storage_account in perm_mapping.prefix:
+                    external_locations.append(perm_mapping)
+        return external_locations
+
+    def _get_tables(self, locations: list[ExternalLocationInfo]):
+        matching_tables = []
+        tables = self._table_mapping.load()
+        for table in tables:
+            table_name = f"{table.catalog_name}.{table.dst_schema}.{table.dst_table}"
+            table_info = self._ws.tables.get(table_name)
+            for location in locations:
+                assert location.url is not None
+                if table_info.storage_location is not None and table_info.storage_location.startswith(location.url):
+                    matching_tables.append(location)
+        return matching_tables
+
+    def _get_external_location(self, locations: list[StoragePermissionMapping]) -> list[ExternalLocationInfo]:
+        matching_location = []
+        for location in self._ws.external_locations.list():
+            for principal_prefix in locations:
+                assert location.url is not None
+                if location.url.startswith(principal_prefix.prefix):
+                    matching_location.append(location)
+        return matching_location
+
+    def _apply_uc_permission(
+        self,
+        locations: list[StoragePermissionMapping],
+    ):
+        matching_external_locations = self._get_external_location(locations)
+        matching_tables = self._get_tables(matching_external_locations)
+        print(matching_tables)
+
+    def _get_cluster_principal_mapping(self, cluster_id: str) -> list[IAMPrincipal]:
+        # gets all the users,groups,spn which have access to the clusters and returns a dataclass of that mapping
+        principal_list = []
+        cluster_permission = self._ws.permissions.get("clusters", cluster_id)
+        if cluster_permission.access_control_list is None:
+            return []
+        for acl in cluster_permission.access_control_list:
+            if acl.user_name is not None:
+                principal_list.append(IAMPrincipal("user", acl.user_name))
```

Comment:
You just need a principal name and not the type - see changes field in https://docs.databricks.com/api/workspace/grants/update